packages/target-tauri/src-tauri/src/util/fs_watcher.rs

```diff
@@ -0,0 +1,66 @@
+use std::{
+    future::Future,
+    path::Path,
+    time::{Duration, SystemTime},
+};
+
+use notify::{Config, Event, RecommendedWatcher, RecursiveMode, Watcher};
+use tauri::async_runtime::{block_on, channel, Receiver};
+
+fn async_watcher() -> notify::Result<(RecommendedWatcher, Receiver<notify::Result<Event>>)> {
+    let (tx, rx) = channel(1);
+
+    let watcher = RecommendedWatcher::new(
+        move |res| {
+            block_on(async {
+                tx.send(res).await.unwrap();
+            })
+        },
+        Config::default(),
+    )?;
+
+    Ok((watcher, rx))
+}
+
+pub async fn async_watch_debounced<P, F>(
+    path: P,
+    callback: Box<dyn Fn() -> F + Sync + Send>,
+    delay: Duration,
+) -> notify::Result<()>
+where
+    P: AsRef<Path>,
+    F: Future<Output = ()>,
+{
+    let (mut watcher, mut rx) = async_watcher()?;
+
+    // Add a path to be watched. All files and directories at that path and
+    // below will be monitored for changes.
+    watcher.watch(path.as_ref(), RecursiveMode::Recursive)?;
+
+    let mut last_action = SystemTime::now();
+
+    while let Some(res) = rx.recv().await {
+        match res {
+            Ok(event) => {
+                log::trace!("changed: {:?}", event);
+                if event.kind.is_create() || event.kind.is_modify() || event.kind.is_remove() {
+                    let now = SystemTime::now();
+                    if let Ok(elapsed) = now.duration_since(last_action) {
+                        if elapsed < delay {
+                            continue;
+                        }
+                        log::trace!("callback triggered");
+                        callback().await;
+                        last_action = now;
+                    } else {
+                        log::error!("debounce: system time error: duration_since");
+                        return Ok(());
+                    }
+                }
+            }
+            Err(e) => log::error!("watch error: {:?}", e),
+        }
```

Comment:
I'd like to remind about the early return pattern (just remind and not force it).

Reply:
how would you do that here?

Reply:
```rust
let event = match res {
    Err(e) => {
        log...
        contunue;
    },
    Ok(event) => event,
};
```

Also `if` can be un-nested.

```rust
if !(,..) {
    continue;
}
```